src/shared_library.c

```diff
@@ -0,0 +1,158 @@
+
+  lib->library_path = rcutils_strdup(library_path, lib->allocator);
+  if (NULL == lib->library_path) {
+    RCUTILS_SET_ERROR_MSG("unable to allocate memory");
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+
+#ifndef _WIN32
+  lib->lib_pointer = dlopen(lib->library_path, RTLD_LAZY);
+  if (!lib->lib_pointer) {
+    lib->allocator.deallocate(lib->library_path, lib->allocator.state);
+    lib->library_path = NULL;
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("dlopen error: %s", dlerror());
+    return RCUTILS_RET_ERROR;
+  }
+#else
+  lib->lib_pointer = LoadLibrary(lib->library_path);
+  if (!lib->lib_pointer) {
+    lib->allocator.deallocate(lib->library_path, lib->allocator.state);
+    lib->library_path = NULL;
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("LoadLibrary error: %lu", GetLastError());
+    return RCUTILS_RET_ERROR;
+  }
+#endif  // _WIN32
+  return RCUTILS_RET_OK;
+}
+
+void *
+rcutils_get_symbol(const rcutils_shared_library_t * lib, const char * symbol_name)
+{
+  if (!lib || !lib->lib_pointer || (symbol_name == NULL)) {
+    RCUTILS_SET_ERROR_MSG("invalid inputs arguments");
+    return NULL;
+  }
+
+#ifndef _WIN32
+  void * lib_symbol = dlsym(lib->lib_pointer, symbol_name);
+  char * error = dlerror();
+  if (error != NULL) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting the symbol '%s'. Error '%s'",
+      symbol_name, error);
+    return NULL;
+  }
+#else
+  void * lib_symbol = GetProcAddress(lib->lib_pointer, symbol_name);
+  if (lib_symbol == NULL) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting the symbol '%s'. Error '%d'",
+      symbol_name, GetLastError());
+    return NULL;
+  }
+#endif  // _WIN32
+  if (!lib_symbol) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "symbol '%s' doesnt' exit in library '%s'",
+      symbol_name, lib->library_path);
+    return NULL;
+  }
+  return lib_symbol;
+}
+
+bool
+rcutils_has_symbol(const rcutils_shared_library_t * lib, const char * symbol_name)
+{
+  if (!lib || !lib->lib_pointer || symbol_name == NULL) {
+    return false;
+  }
+
+#ifndef _WIN32
+  // the correct way to test for an error is to call dlerror() to clear any old error conditions,
+  // then call dlsym(), and then call dlerror() again, saving its return value into a variable,
+  // and check whether this saved value is not NULL.
+  dlerror(); /* Clear any existing error */
+  void * lib_symbol = dlsym(lib->lib_pointer, symbol_name);
+  return dlerror() == NULL && lib_symbol != 0;
+#else
+  void * lib_symbol = GetProcAddress(lib->lib_pointer, symbol_name);
+  return GetLastError() == 0 && lib_symbol != 0;
+#endif  // _WIN32
+}
+
+rcutils_ret_t
+rcutils_unload_shared_library(rcutils_shared_library_t * lib)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(lib, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(lib->lib_pointer, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(lib->library_path, RCUTILS_RET_INVALID_ARGUMENT);
```

Comment:
Since the function later uses `lib->allocator` it should check that it isn't empty.